### ISSUE TITLE
Associated records counted incorrectly when shown concatenated on same row (0.4.0)

### ISSUE THREAD

**Issue by ghost** [jbox-web/ajax-datatables-rails#171]
Consider the model `Event` with a `has_many :people, through: attendees` association and the following test data:

```
Event1: {id: 1, name: "Event1", date: "2016-08-08"}
Event2: {id: 2, name: "Event2", date: "2016-07-07"}
Person1: {id: 1, name: "Jack"}
Person2: {id: 2, name: "Jill"}
Person3: {id: 3, name: "Pete"}
Attendees:
{event_id: 1, person_id: 1}
{event_id: 1, person_id: 2}
{event_id: 2, person_id: 2}
{event_id: 2, person_id: 3}
```

(Both events were attended by 2 people, Jill attended both events)

What I would like to display using this gem:

| Event | Date | Attendees |
| --- | --- | --- |
| Event1 | 2016-08-08 | Jack, Jill |
| Event2 | 2016-07-07 | Jill, Pete |

I have set up the following:

```
  def view_columns
    @view_columns ||= {
        event: { source: 'Event.name', cond: :like },
        date: { source: 'Event.date', cond: :eq },
        people: { source: 'Person.name' }
    }
  end

  def data
    records.map do |event|
      {
        event: event.name,
        date: l(event.date),
        people: event.people.map {|a| a.name}.join('<br/>')
      }
  end

  def get_raw_records
    Event.references(:people).includes(:people).distinct
  end
```

This will show 2 rows correctly like I expected, with all attendees concatenated on the same line.

However, something goes wrong with the counts:
`Showing 1 to 25 of 4 entries (filtered from 2 total entries)`

I have tracked this problem to the following method in `base.rb` of the gem.

```
    def as_json(options = {})
      {
        recordsTotal: get_raw_records.count(:all),
        recordsFiltered: get_raw_records.model.from("(#{filter_records(get_raw_records).except(:limit, :offset, :order).to_sql}) AS foo").count,
        data: data
      }
    end
```

`recordsTotal` generates a correct disintct statement as follows:

```
SELECT DISTINCT COUNT(DISTINCT `events`.`id`) FROM `events` LEFT OUTER JOIN `attendees` ON ...
```

while `recordsFiltered` gets an incorrect count:

```
SELECT COUNT(*) FROM (SELECT DISTINCT `events`.`id` AS t0_r0, `events`.`name` AS t0_r1 ... ) AS foo
```

Any idea on how to correct this issue? This is with the 0.4.0 version.

**Comment by ajahongir** [comment]
`recordsTotal` generates this query and 

``` sql
SELECT DISTINCT COUNT(DISTINCT `events`.`id`) FROM `events` LEFT OUTER JOIN `attendees` ON ...
```

returns 4?

**Comment by ghost** [comment]
No. That returns 2 correctly, since it takes takes the count of distinct event id's (2 total in 4 rows with the joins)

That other one in `recordsFiltered` returns 4 as it takes the count of distinct over all columns instead of just `events.id`

**Comment by zenzei** [comment]
It's because the left join  (subquery) generates more than one record for each main model (in your case event) if the has_many relation has more than one element (in your case person).

I also had this problem, and fix it adding a distinct at the end, so instead of this

```
recordsFiltered: get_raw_records.model.from("(#{filter_records(get_raw_records).except(:limit, :offset, :order).to_sql}) AS foo").count
```

I have now this

```
recordsFiltered: get_raw_records.model.from("(#{filter_records(get_raw_records).except(:limit, :offset, :order).to_sql}) AS foo").distinct.count('t0_r0')
```

I'm not proud with this ugly solution but it works in my use case. But note that if you use a _select_ clause in the _get_raw_records_ it's highly possible that this will not work.

**Comment by ghost** [comment]
Is there a downside to your solution?

Is there a case where this breaks intended functionality?

**Comment by zenzei** [comment]
As i understand, if you don't use a "select" clause in the _get_raw_records_ there should be no break in functionality, as the "distinct count" is done over _t0_r0_, the first column (id) of the reference model.

**Comment by zenzei** [comment]
I was trying a few more things today and see that if you replace _recordsFiltered_ with the next line, the filtered count query also works as expected (and should also work if the get_raw_records has a _select_ clause)

```
recordsFiltered: filter_records(get_raw_records).except(:limit, :offset, :order).count,
```

It seems much clear in this way, but i'm not sure why @antillas21 / @ajahongir (i think one of them are the original authors) wrapped this query as a subquery. So maybe we should wait for their confirmation.

**Comment by ajahongir** [comment]
well, It was a reason to wrap query as sub...
I am not sure that we can write single query for all cases. Current query was build to support nested and joined queries. But I can see that its not cover all cases. So I think we have to do methods overrideable, something like:

``` ruby
def as_json(options = {})
  {
    recordsTotal: total_records(options),
    recordsFiltered: filtered_records(options),
    data: data
  }
end

def total_records options
  get_raw_records.count(:all)
end

def filtered_records options
  get_raw_records.model.from("(#{filter_records(get_raw_records).except(:limit, :offset, :order).to_sql}) AS foo").count
end
```

so you can any time redefine these methods not hole as_json.

**Comment by zenzei** [comment]
It's ok for me, altough I think that both the _total_records_ and the _filtered_records_ default implementation should follow the same behaviour. 

As I see active record do a _COUNT DISTINCT model_id_ (_events.id_ in cdekker example) when doing a _count(:all)_ that have includes/references in the _get_raw_records_ (something that is needed and indeed mark as important in the gem readme). 

So to get the same behaviour in the _filtered_records_ methods i think the _count_ should be applied in the subquery, because if not you will have incorrect results.

Nevetherless I was searching the issue list and I'm not quiet sure why this never happen before as it seems to me that having a has_many relation with more than one element in the relation is a common use case.

By the way, I forgot to mention before but you create a very useful gem that got a lot more powerful in the 0.4.0 branch so thanks a lot for that!!! :+1:  :wink:

**Comment by ajahongir** [comment]
guys, I just want to say that we have some discussion here https://github.com/antillas21/ajax-datatables-rails/pull/105 so if you have any ideas or you can help to release new features or updating readme - It will be just great!

> So to get the same behaviour in the filtered_records methods i think the count should be applied in the subquery, because if not you will have incorrect results.

sure, that's why I did so. But new interface of datatable.rb should be more flexible.

> By the way, I forgot to mention before but you create a very useful gem that got a lot more powerful in the 0.4.0 branch so thanks a lot for that!!!

thanks, I appreciate

**Comment by zenzei** [comment]
> So to get the same behaviour in the filtered_records methods i think the count should be applied in the subquery, because if not you will have incorrect results.
> 
> sure, that's why I did so. But new interface of datatable.rb should be more flexible.

Maybe i wasn't not clear, but I mean that the _count_ should be inside the inner subquery (as my example), and not in the outer query (like current 0.4.0 code), because if not the count would lead to improper results. 

But if you prefer to close this one we can continue in #105!

**Comment by ajahongir** [comment]
current query also works with custom columns

``` ruby
def filtered_records options
  get_raw_records.model.from("(#{filter_records(get_raw_records).except(:limit, :offset, :order).to_sql}) AS foo").count
end
```

like here - https://github.com/ajahongir/ajax-datatables-rails-v-0-4-0-how-to/blob/master/app/datatables/city_datatable.rb#L7

**Comment by zenzei** [comment]
I'm not sure what you mean, do you think that my proposed "count" change can break something with custom columns? Cause right now i have several of those customs working without problems, using the 0.4.0 code with that "count" update.

Nevetherless as I say before (and seeing @cdekker didn't answer again) you can close the issue if you want or we can continue in #105 ;)

**Comment by dpsk** [comment]
Don't want to create separate issue but i noticed that count is taking way too much time using Postgres:

 SQL (62.6ms)  big select with joins and filtering
`SQL (873.5ms)  SELECT  DISTINCT "accounts"."id", accounts.name AS alias_0 FROM "accounts" LEFT OUTER JOIN "account_brands" ON "account_brands"."account_id" = "accounts"."id" LEFT OUTER JOIN "brands" ON "brands"."id" = "account_brands"."brand_id" AND "brands"."is_active" = $1 LEFT OUTER JOIN "addresses" ON "addresses"."addressable_id" = "accounts"."id" AND "addresses"."address_type" = $2 AND "addresses"."addressable_type" = $3 LEFT OUTER JOIN "contacts" ON "contacts"."account_id" = "accounts"."id" LEFT OUTER JOIN "taggings" ON "taggings"."account_id" = "accounts"."id" LEFT OUTER JOIN "tags" ON "tags"."id" = "taggings"."tag_id" WHERE "accounts"."organization_id" = $4 AND "accounts"."is_active" = $5  ORDER BY accounts.name ASC LIMIT 50 OFFSET 0  [["is_active", "t"], ["address_type", "mailing"], ["addressable_type", "Account"], ["organization_id", 32], ["is_active", "t"]]`
`(935.6ms)  SELECT COUNT(DISTINCT "accounts"."id") FROM "accounts" LEFT OUTER JOIN "account_brands" ON "account_brands"."account_id" = "accounts"."id" LEFT OUTER JOIN "brands" ON "brands"."id" = "account_brands"."brand_id" AND "brands"."is_active" = $1 LEFT OUTER JOIN "addresses" ON "addresses"."addressable_id" = "accounts"."id" AND "addresses"."address_type" = $2 AND "addresses"."addressable_type" = $3 LEFT OUTER JOIN "contacts" ON "contacts"."account_id" = "accounts"."id" LEFT OUTER JOIN "taggings" ON "taggings"."account_id" = "accounts"."id" LEFT OUTER JOIN "tags" ON "tags"."id" = "taggings"."tag_id" WHERE "accounts"."organization_id" = $4 AND "accounts"."is_active" = $5  [["is_active", "t"], ["address_type", "mailing"], ["addressable_type", "Account"], ["organization_id", 32], ["is_active", "t"]]`
This is kinda to much almost 1s each, and why is there are 2 of them? Any ideas?

My current version is 0.3.1, a lot of changes in 0.4.0 that will require a lot of rewriting from my side, i also not sure if it's addressed in the new version or  not?

**Comment by ajahongir** [comment]
You are using too many joins in your query

**Comment by dpsk** [comment]
I have pretty intense table, is there any other option other than joins to avoid N+1?
Can i somehow override the gem is counting records? Maybe wrap those count functions as methods and expose them to public?


--
Mikhail Nikalyukin

**Comment by ajahongir** [comment]
Could you put your datatabe here?

**Comment by dpsk** [comment]
Query: `record.accounts.active.includes(:mailing_address, :contacts, taggings: :tag)`
Datatable mapping(showing only related to joins):
```
@searchable_columns ||= ['Account.name','Contact.first_name','Contact.title','Contact.phone_no','Contact.fax', 'Address.address_line1', 'Account.general_notes', 'Account.website', 'Contact.email','Address.city','Address.state','Address.zip_code','Brand.name', 'Account.joor_id']

    records.map do |account|
      [
        account_link(account, list), #account taggings with tag name shown for each link
        account.contacts.map(&:name).join('<br/>').html_safe,
        account.mailing_address.try(:show_address),
        account.active_brands.map{|brand| link_to brand.name, brand_path(brand),
      ]
    end
```

**Comment by ajahongir** [comment]
well, you can override here count method BUT joinug more than 3-4 table is always too slow. you have to change your architec.

```ruby
    def as_json(options = {})
      {
        :draw => params[:draw].to_i,
        :recordsTotal =>  get_raw_records.count(:all),
        :recordsFiltered => filter_records(get_raw_records).count(:all),
        :data => data
      }
    end
```

**Comment by zenzei** [comment]
Hi @ajahongir, as I see in the last released version (v0.4.0) you already remove the subquery for the recordsFiltered count as i suggest in the previous comments, so i think this issue can be closed now :)
